lib/tbot/service/supervisor.go

```diff
@@ -0,0 +1,258 @@
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package service
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/tbot/service/status"
+)
+
+// backoffResetPeriod is the amount of time a service's handler needs to run
+// before exiting for us to consider it "stable" and reset the retry backoff.
+//
+// Resetting the backoff avoids a potential issue where a service fails many
+// times in quick succession (e.g. during a network partition), stabilizes, but
+// later fails again and is penalized for its earlier failures.
+const backoffResetPeriod = 10 * time.Minute
+
+// Supervisor is responsible for managing the lifecycle of tbot's component
+// services.
+type Supervisor struct {
+	logger *slog.Logger
+	clock  clockwork.Clock
+
+	mu       sync.Mutex
+	services map[string]InternalService
+	started  bool
+}
+
+// SupervisorConfig contains the configuration options for Supervisor.
+type SupervisorConfig struct {
+	// Logger used to log messages and errors.
+	Logger *slog.Logger
+
+	// Clock allows you to override the clock in unit tests.
+	Clock clockwork.Clock
+}
+
+func (cfg *SupervisorConfig) checkAndSetDefaults() error {
+	if cfg.Logger == nil {
+		cfg.Logger = slog.Default()
+	}
+	if cfg.Clock == nil {
+		cfg.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// NewSupervisor creates a new supervisor.
+func NewSupervisor(cfg SupervisorConfig) (*Supervisor, error) {
+	if err := cfg.checkAndSetDefaults(); err != nil {
+		return nil, err
+	}
+	return &Supervisor{
+		services: make(map[string]InternalService),
+		logger:   cfg.Logger,
+		clock:    cfg.Clock,
+	}, nil
+}
+
+// Register the given service. Returns an error if the service has already been
+// registered, or if the supervisor is already running.
+//
+// TODO: support registering services dynamically at-runtime.
+func (s *Supervisor) Register(service InternalService) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.started {
+		return trace.Errorf("cannot register services once the supervisor has started")
+	}
+
+	if _, ok := s.services[service.getName()]; ok {
+		return trace.Errorf("service %q is already registered", service.getName())
```

Comment:
This is nitpicky but the real check if the service is registered is below. This only checks for the name collision.

```suggestion
		return trace.Errorf("service with name %q is already registered", service.getName())
```